File: 2188-minimized-maximum-of-products-distributed-to-any-store/2188-minimized-maximum-of-products-distributed-to-any-store.cpp
class Solution {
public:
    bool isPossileDist(int x, vector<int>& quantities, int shops) {
        for(int &products : quantities)
        {
            shops -= (products + x - 1)/x;
            if(shops < 0) return false;
        }
        return true;
    }

    int minimizedMaximum(int n, vector<int>& quantities) {
        int m = quantities.size(), maxi = quantities[0];
        for(int i = 0; i < m; i++)
        {
            maxi = max(maxi, quantities[i]);
        }
        int l = 1, r = maxi, ans = 0;;
        while(l <= r)
        {
            int mid = l + (r - l) / 2;
            if(isPossileDist(mid, quantities, n))
            {
                ans = mid;
                r = mid - 1;
            }
            else l = mid + 1;
        }
        return ans;
    }
};

File: 2188-minimized-maximum-of-products-distributed-to-any-store/2188-minimized-maximum-of-products-distributed-to-any-store.java
class Solution {
    public int minimizedMaximum(int n, int[] q) {
        int s = 1, e = Integer.MIN_VALUE;

        for(int i = 0; i < q.length; i++)
        {
            if(q[i] > e) e = q[i];
        }

        int ans = -1;

        while(s <= e)
        {
            int mid = s + (e - s)/2;
            if(isPossible(q, mid, n))
            {
                ans = mid;
                e = mid - 1;
            }
            else s = mid + 1;
        }
        return ans;
    }

    boolean isPossible(int[] q, int mid, int store)
    {
        int sum = 0;

        for(int i = 0; i < q.length; i++)
        {
            sum += q[i] / mid;

            if(q[i] % mid != 0) sum++;

            if(sum > store) return false;
        }

        return true;
    }
}